### PR TITLE
Add NCS example to showcase LLPM features

### PR DESCRIPTION
**Pull request by ryanjh**
As a user of the ble-controller_0.3.0 with its new and shiny LLPM features, they may have trouble understanding what LLPM actually is. It would be super-nice to have an example showing the users how these features are used in practice.

### CODE REVIEW

**Comment by ryanjh** [comment]
@carlescufi Would you please help to invite 'Ferenc, Grzegorz' as the reviewer?

File: include/bluetooth/services/latency.h [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+			    0x9B, 0xF3, 0xDB, 0x58, 0x01, 0x6E, 0x13, 0x67)
+
+/** @brief UUID of the Latency Characteristic. **/
+#define BT_UUID_LATENCY_CHAR                                                \
+	BT_UUID_DECLARE_128(0x3A, 0x81, 0xC0, 0x58, 0xDE, 0xFD, 0x46, 0x34, \
+			    0x9B, 0xF3, 0xDB, 0x58, 0x02, 0x6E, 0x13, 0x67)
+
+/** @brief Initialize the GATT Latency Service.
+ *
+ *  @param[in] Latency service instance.
```
Comment (greg-fer) [inline include/bluetooth/services/latency.h]: This will be rendered with `Latency` as parameter name. Please provide parameter name. (Mirror this edit elsewhere in this file.)

File: include/bluetooth/services/latency.h [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+ *  @param[in] cb Callbacks.
+ *
+ *  @retval 0 If the operation was successful.
+ *            Otherwise, a negative error code is returned.
```
Comment (greg-fer) [inline include/bluetooth/services/latency.h]: Please use two `retval` here. One for `0`, one for `error`. (Mirror this edit elsewhere in this file.)

File: include/bluetooth/services/latency.h [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+
+	/** Connection object. */
+	struct bt_conn *conn;
+};
+
+/** @brief Latency Service UUID. */
+#define LATENCY_UUID_DATA                                           \
+	((u8_t []) {0x3A, 0x81, 0xC0, 0x58, 0xDE, 0xFD, 0x46, 0x34, \
+		    0x9B, 0xF3, 0xDB, 0x58, 0x01, 0x6E, 0x13, 0x67})
+#define BT_UUID_LATENCY                                                     \
```
Comment (ryanjh) [inline include/bluetooth/services/latency.h]: Based on the discussion with @rakons , all the samples should migrate to the UUID-128 encoder which has not been merged to fw-nrfconnect-nrf yet. I will append one more change for it later on.
https://github.com/zephyrproject-rtos/zephyr/pull/19821/commits/460502ac862cca5503dae51d7a9961db410945b3

@carlescufi Do we have a planed downmerge?

File: samples/bluetooth/llpm/src/main.c [hunk 4]
```diff
@@ -0,0 +1,412 @@
...
+		default_conn = NULL;
+	}
+	advertise_and_scan();
+}
+
+static void le_param_updated(struct bt_conn *conn, u16_t interval,
+			     u16_t latency, u16_t timeout)
+{
+	if (interval == INTERVAL_LLPM) {
+		printk("Connection interval updated: LLPM (1 ms)\n");
```
Comment (KAGA164) [inline samples/bluetooth/llpm/src/main.c]: What about a situation, when different interval than 1 ms was set. We do not need to inform about it ?

**Comment by KAGA164** [comment]
Generally, is conceptual documentation for latency service missing ?

File: include/bluetooth/services/latency.h [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+/** @brief Initialize the GATT Latency Service.
+ *
+ *  @param[in] Latency service instance.
+ *  @param[in] cb Callbacks.
+ *
+ *  @retval 0 If the operation was successful.
+ *            Otherwise, a negative error code is returned.
+ */
+int bt_gatt_latency_init(struct bt_gatt_latency *latency,
+			 const struct bt_gatt_latency_cb *cb);
```
[the remark this answers is not in the document]
Comment (ryanjh) [reply include/bluetooth/services/latency.h]: I checked it by "View file". It should be aligned. Isn't it? Is Zephyr using 8 spaces for a tab?

https://github.com/NordicPlayground/fw-nrfconnect-nrf/blob/9817040dccd8a1109ab041e237c25cc45b08e9f9/include/bluetooth/services/latency.h

File: include/bluetooth/services/latency.h [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+	atomic_t state;
+};
+
+/** @brief Latency Service UUID. */
+#define LATENCY_UUID_DATA                                           \
+	((u8_t []) {0x3A, 0x81, 0xC0, 0x58, 0xDE, 0xFD, 0x46, 0x34, \
+		    0x9B, 0xF3, 0xDB, 0x58, 0x01, 0x6E, 0x13, 0x67})
+#define BT_UUID_LATENCY                                                     \
+	BT_UUID_DECLARE_128(0x3A, 0x81, 0xC0, 0x58, 0xDE, 0xFD, 0x46, 0x34, \
+			    0x9B, 0xF3, 0xDB, 0x58, 0x01, 0x6E, 0x13, 0x67)
```
[the remark this answers is not in the document]
Comment (ryanjh) [reply include/bluetooth/services/latency.h]: Yup, I had the same idea at the beginning, but it doesn't get approved by the coding-style of Zephyr.

-:9: WARNING:COMPLEX_MACRO: Macros with complex values should be enclosed in parentheses
#9: FILE: include/bluetooth/services/latency.h:60:
+#define LATENCY_UUID_TEST 0x23, 0xD1, 0xBC, 0xEA, 0x5F, 0x78, 0x23, 0x15, \
+                         0xDE, 0xEF, 0x12, 0x12, 0x23, 0x15, 0x00, 0x00

- total: 0 errors, 1 warnings, 9 lines checked

File: samples/bluetooth/llpm/src/main.c (hunk 4 above)
Comment (ryanjh) [reply samples/bluetooth/llpm/src/main.c]: In this sample, the le_param_update is only used to enable the llpm interval which is fixed to 1 ms. Moreover, this LLPM interval is a predefined value and it is far from the definition in BLE spec

File: include/bluetooth/services/latency.h (hunk 5 above)
[the remark this answers is not in the document]
Comment (joerchan) [reply include/bluetooth/services/latency.h]: Sometimes it looks a bit unaligned. I think it is the '+' character eating up some of the tab-width. It will look the same in 'vim' for example.

File: include/bluetooth/services/latency.h (hunk 3 above)
Comment (joerchan) [reply include/bluetooth/services/latency.h]: I think it will be next week. Someone will go over and update all instances afterwards.

File: include/bluetooth/services/latency.rst [hunk 7]
```diff
@@ -0,0 +1,38 @@
+.. _latency_readme:
+
+GATT Latency Service
+####################
+
+The GATT Latency Service is a custom service that receives data and returns result about these operations.
```
Comment (greg-fer) [inline include/bluetooth/services/latency.rst]:
```suggestion
The GATT Latency Service is a custom service that receives data and returns result about GATT latency operations.
```
`These` here is ambiguous, as you don't mention explicitly any operation before.

File: include/bluetooth/services/latency.rst [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+
+To measure GATT transmission latency, when a client writes a timestamp to the latency characteristic on the server, the server will automatically reply a response back.
```
Comment (greg-fer) [inline include/bluetooth/services/latency.rst]:
```suggestion
To measure the GATT transmission latency, when a client writes a timestamp to the latency characteristic on the server, the server will automatically reply back.
```

File: subsys/bluetooth/services/latency.c [hunk 9]
```diff
@@ -0,0 +1,159 @@
+/*
```
Comment (joerchan) [inline subsys/bluetooth/services/latency.c]: This seems to be both Service and Client in one file. We usually have one file for each. I'd suggest to split it.

File: include/bluetooth/services/latency.h (hunk 6 above)
[the remark this answers is not in the document]
Comment (KAGA164) [reply include/bluetooth/services/latency.h]: Fortunately, warnings are not block a merge. Sometimes, for greater code readability, it is worth ignoring. I think in this place you can simplify this defines to one.

File: include/bluetooth/services/latency.h (hunk 5 above)
[the remark this answers is not in the document]
Comment (KAGA164) [reply include/bluetooth/services/latency.h]: Ok. I see now.

File: include/bluetooth/services/latency.h (hunk 6 above)
[the remark this answers is not in the document]
Comment (KAGA164) [reply include/bluetooth/services/latency.h]: And more over the latency service will be consistent with our other custom service. You can check the lbs and the uart service.

File: samples/bluetooth/llpm/src/main.c (hunk 4 above)
Comment (KAGA164) [reply samples/bluetooth/llpm/src/main.c]: So is not possible to get for some reason here at other value of interval which is invalid for this sample ?

File: include/bluetooth/services/latency.h (hunk 6 above)
[the remark this answers is not in the document]
Comment (joerchan) [reply include/bluetooth/services/latency.h]: Being consistent with the rest of the services is a good idea. The warning can be ignored.

File: subsys/bluetooth/services/Kconfig.latency [hunk 10]
```diff
@@ -0,0 +1,19 @@
+#
+# Copyright (c) 2019 Nordic Semiconductor ASA
+#
+# SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+#
+
+config BT_GATT_LATENCY
+	bool "Latency service"
+	default n
```
[the remark this answers is not in the document]
Comment (ryanjh) [reply subsys/bluetooth/services/Kconfig.latency]: ok, then BT_GATT_CLIENT can be moved also.

File: samples/bluetooth/llpm/README.rst [hunk 11]
```diff
@@ -0,0 +1,313 @@
...
+
+The Bluetooth Low Latency Packet Mode (LLPM) sample uses the :ref:`latency_readme` to showcase the LLPM proprietary Bluetooth extension from Nordic Semiconductor.
+You can use it to determine the transmission latency of LLPM-enabled connections, or to compare with different connection parameters and check their influence on the results.
+
+
+Overview
+********
+
+The LLPM is designed time-sensitive transmissions, in which time synchronization is important and interface response time is critical for the user.
+For example, for reality headsets or gaming mouse and keyboard.
```
[the remark this answers is not in the document]
Comment (ryanjh) [reply samples/bluetooth/llpm/README.rst]: @greg-fer How do you think?

[the remark this answers is not in the document]
Comment (joerchan) [reply samples/bluetooth/llpm/README.rst]: I'm just asking. Maybe I just don't know what reality headset are 😛

File: include/bluetooth/services/latency.h (hunk 6 above)
[the remark this answers is not in the document]
Comment (ryanjh) [reply include/bluetooth/services/latency.h]: I managed to resolve the coding-stype problem with a magic macro in util.h as below. It would look better now. :)

#define LATENCY_UUID                                                \
	UTIL_EXPAND(0x3A, 0x81, 0xC0, 0x58, 0xDE, 0xFD, 0x46, 0x34, \
		    0x9B, 0xF3, 0xDB, 0x58, 0x01, 0x6E, 0x13, 0x67)

#define BT_UUID_LATENCY BT_UUID_DECLARE_128(LATENCY_UUID)

It can also be used in the example (main.c).

static const struct bt_data ad[] = {
	BT_DATA_BYTES(BT_DATA_FLAGS, (BT_LE_AD_GENERAL | BT_LE_AD_NO_BREDR)),
	BT_DATA_BYTES(BT_DATA_UUID128_ALL, LATENCY_UUID),
};

File: samples/bluetooth/llpm/src/main.c (hunk 4 above)
Comment (ryanjh) [reply samples/bluetooth/llpm/src/main.c]: Per discussion, this sample sets INTERVAL_LLPM to both interval_min and interval_max, it won't get a callback if the interval is unnegotiable with its peer.

File: subsys/bluetooth/services/latency.c (hunk 9 above)
Comment (ryanjh) [reply subsys/bluetooth/services/latency.c]: Because there is only one special case in function test_run() on the master, we tend to keep it simple as same as throughput. Then, it won't introduce too much duplicate code,

File: include/bluetooth/services/latency.rst [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+
+GATT Latency Service
+####################
+
+The GATT Latency Service is a custom service that receives data and returns result about GATT latency operations.
+
+To measure the GATT transmission latency, when a client writes a timestamp to the latency characteristic on the server, the server will automatically reply back.
+The client can then use its present time and the corresponding timestamp written before to estimate the round-trip time (RTT).
+
+The GATT Latency Service is used in the :ref:`ble_llpm` sample.
```
Comment (rugeGerritsen) [inline include/bluetooth/services/latency.rst]: We should probably not have references to examples using the service as this may get outdated

File: include/bluetooth/services/latency.rst [hunk 13]
```diff
@@ -0,0 +1,38 @@
+.. _latency_readme:
+
+GATT Latency Service
+####################
+
+The GATT Latency Service is a custom service that receives data and returns result about GATT latency operations.
```
Comment (rugeGerritsen) [inline include/bluetooth/services/latency.rst]: Would rephrase it to:
"The GATT Latency Service is a custom service containing a single writable characteristic. This characteristic may be used to calculate the latency of a GATT Write operation"

File: include/bluetooth/services/latency.rst [hunk 14]
```diff
@@ -0,0 +1,38 @@
...
+
+To measure the GATT transmission latency, when a client writes a timestamp to the latency characteristic on the server, the server will automatically reply back.
```
Comment (rugeGerritsen) [inline include/bluetooth/services/latency.rst]: I believe we can remove the paragraph describing how to measure the latency, as this is done in the application, not in the service itself

File: samples/bluetooth/llpm/README.rst [hunk 15]
```diff
@@ -0,0 +1,313 @@
...
+
+See the following subsections for a description of the key LLPM elements.
+
+LLPM connection interval (1 ms)
+   The connection interval defines how often the devices must listen on the radio.
+   The LLPM introduces the possibility to reduce the connection interval below what is supported in BLE.
+   The lowest interval supported is 1 ms for one link.
```
Comment (rugeGerritsen) [inline samples/bluetooth/llpm/README.rst]: "The lowest supported connection interval is 1 ms"

File: include/bluetooth/services/latency.rst (hunk 12 above)
Comment (greg-fer) [reply include/bluetooth/services/latency.rst]: I agree. Let's rather have a link from the example to the service.

File: samples/bluetooth/llpm/README.rst (hunk 15 above)
Comment (greg-fer) [reply samples/bluetooth/llpm/README.rst]: I agree. Let's rather have a link from the example to the service.

File: subsys/bluetooth/services/latency.c (hunk 9 above)
Comment (ryanjh) [reply subsys/bluetooth/services/latency.c]: Alright, I misunderstood it in the first place. It sounds good to split the latency service to be reusable.

File: samples/bluetooth/llpm/README.rst [hunk 16]
```diff
@@ -0,0 +1,313 @@
...
+###############
+
+The Bluetooth Low Latency Packet Mode (LLPM) sample uses the :ref:`latency_readme` to showcase the LLPM proprietary Bluetooth extension from Nordic Semiconductor.
+You can use it to determine the transmission latency of LLPM-enabled connections, or to compare with different connection parameters and check their influence on the results.
+
+
+Overview
+********
+
+The LLPM is designed time-sensitive transmissions, in which time synchronization is important and interface response time is critical for the user.
```
Comment (rugeGerritsen) [inline samples/bluetooth/llpm/README.rst]: Would rephrase to: "LLPM is designed for applications where interface response time is critical for the user". Time synchronization is not one of the main use cases

File: samples/bluetooth/llpm/README.rst [hunk 17]
```diff
@@ -0,0 +1,313 @@
+.. _ble_llpm:
+
+Bluetooth: LLPM
+###############
+
+The Bluetooth Low Latency Packet Mode (LLPM) sample uses the :ref:`latency_readme` to showcase the LLPM proprietary Bluetooth extension from Nordic Semiconductor.
```
Comment (rugeGerritsen) [inline samples/bluetooth/llpm/README.rst]: Would rephrase it to "The Bluetooth Low Latency Packet Mode (LLPM) sample uses the :ref:`throughput_readme` to measure the GATT Write latency. It uses the properietary LLPM Bluetooth extensions to achieve a latency below 7.5 ms. The sample can also be used to determine how different connection parameters affect the latency"

File: samples/bluetooth/llpm/README.rst (hunk 11 above)
[the remark this answers is not in the document]
Comment (greg-fer) [reply samples/bluetooth/llpm/README.rst]:
```suggestion
For example, for virtual reality headsets or gaming mouse and keyboard.
```
Do you mean virtual reality headsets? I understood this was the meaning.

File: samples/bluetooth/llpm/README.rst (hunk 16 above)
Comment (greg-fer) [inline samples/bluetooth/llpm/README.rst]:
```suggestion
The LLPM is designed for applications in which the interface response time is critical for the user.
```

File: samples/bluetooth/llpm/src/main.c [hunk 18]
```diff
@@ -0,0 +1,412 @@
...
+		printk("Press any key to set LLPM short connection interval (1 ms)\n");
+		console_getchar();
+
+		if (enable_llpm_short_connection_interval()) {
+			printk("Enable LLPM short connection interval failed\n");
+			return;
+		}
+	}
+
+	printk("Press any key to start measuring transmission latency\n");
```
[the remark this answers is not in the document]
Comment (rugeGerritsen) [reply samples/bluetooth/llpm/src/main.c]: It is probably better to keep it as is, and use a key, to align with other examples

File: samples/bluetooth/llpm/README.rst [hunk 19]
```diff
@@ -0,0 +1,313 @@
...
+   - The master outputs the following information::
+
+       Press any key to set LLPM short connection connection interval (1 ms)
+
+   - The slave outputs the following information::
+
+       Press any key to start measuring transmission latency
+
+#. Press a key in the terminal that is connected to the slave.
+#. Observe that the slave starts sending data to the master and measuring the transmission latency with the default connection interval::
```
Comment (rugeGerritsen) [inline samples/bluetooth/llpm/README.rst]: Would rephrase to:
```suggestion
#. Observe that latency measurements are printed in the terminal. The latency is expected to be slightly shorter than the connection interval:
```

File: samples/bluetooth/llpm/README.rst [hunk 20]
```diff
@@ -0,0 +1,313 @@
...
+
+#. Press a key in the terminal that is connected to the master.
+#. Observe the output on the master.
+   It is waiting for a user input::
+
+       Press any key to start measuring transmission latency
+
+#. Observe the connection gets updated to LLPM connection interval (1 ms) on both sides::
+
+       Connection intervonal updated: LLPM (1 ms)
```
Comment (rugeGerritsen) [inline samples/bluetooth/llpm/README.rst]: Typo:
```suggestion
       Connection interval updated: LLPM (1 ms)
```

File: samples/bluetooth/llpm/README.rst (hunk 19 above)
Comment (greg-fer) [reply samples/bluetooth/llpm/README.rst]: As per RST formatting, the colon needs to be double `::`. Otherwise, the RST will treat the single colon as a linking markup and break the formatting.

File: samples/bluetooth/llpm/README.rst [hunk 21]
```diff
@@ -0,0 +1,313 @@
...
+
+#. Observe the output on the slave.
+   While the slave keeps sending data to the master with the LLPM connection interval, the measured latency on the slave becomes approximate 1 ms::
```
Comment (rugeGerritsen) [inline samples/bluetooth/llpm/README.rst]: Suggested rephrasing: 
"On the terminal connected to the slave, observe the latency is reduced to approximately 1 ms."

File: samples/bluetooth/llpm/README.rst [hunk 22]
```diff
@@ -0,0 +1,313 @@
...
+
+       Transmission Latency: 1098 (us)
+
+#. Press a key in the terminal that is connected to the master.
+#. Observe the output on the master.
```
Comment (rugeGerritsen) [inline samples/bluetooth/llpm/README.rst]: I would simplify this phrasing to be similar to the slave.
"On the terminal connected to the master, observe the latency is approximately 1 ms."

File: samples/bluetooth/llpm/src/main.c [hunk 23]
```diff
@@ -0,0 +1,415 @@
...
+}
+
+void scan_filter_no_match(struct bt_scan_device_info *device_info,
+			  bool connectable)
+{
+	char addr[BT_ADDR_LE_STR_LEN];
+
+	bt_addr_le_to_str(device_info->addr, addr, sizeof(addr));
+
+	printk("Filter not match. Address: %s connectable: %d\n",
```
Comment (rugeGerritsen) [inline samples/bluetooth/llpm/src/main.c]:
```suggestion
	printk("Filter does not match. Address: %s connectable: %d\n",
```

File: samples/bluetooth/llpm/README.rst [hunk 24]
```diff
@@ -0,0 +1,313 @@
...
+
+  * :file:`include/bluetooth/bluetooth.h`
+  * :file:`include/bluetooth/conn.h`
+  * :file:`include/bluetooth/gatt.h`
+  * :file:`include/bluetooth/hci.h`
+  * :file:`include/bluetooth/uuid.h`
+  * :file:`include/bluetooth/scan.h`
+  * :file:`include/bluetooth/gatt_dm.h`
+
+Moreover, it uses the following Nrfx libraries:
```
Comment (rugeGerritsen) [inline samples/bluetooth/llpm/README.rst]: @greg-fer  , how do we document this properly? "It uses the BLE controller library in nrfxlib"?

File: samples/bluetooth/llpm/README.rst [hunk 25]
```diff
@@ -0,0 +1,313 @@
...
+References
+***********
+
+For more information about the connection parameters that are used in this sample, see the following chapters of the `Bluetooth Core Specification`_:
+
+* Vol 3, Part F, 3.2.8 Exchanging MTU Size
+* Vol 3, Part F, 3.4.5 Writing attributes
+* Vol 3, Part G, 4.9.3 Write Characteristic Value
+* Vol 6, Part B, 5.1.1 Connection Update Procedure
+* Vol 6, Part B, 5.1.9 Data Length Update Procedure
```
Comment (rugeGerritsen) [inline samples/bluetooth/llpm/README.rst]:
```suggestion

```

File: samples/bluetooth/llpm/README.rst [hunk 26]
```diff
@@ -0,0 +1,313 @@
...
+
+  * :file:`include/ble_controller_hci_vs.h`
+
+
+References
+***********
+
+For more information about the connection parameters that are used in this sample, see the following chapters of the `Bluetooth Core Specification`_:
+
+* Vol 3, Part F, 3.2.8 Exchanging MTU Size
```
Comment (rugeGerritsen) [inline samples/bluetooth/llpm/README.rst]:
```suggestion
```

File: samples/bluetooth/llpm/src/main.c [hunk 27]
```diff
@@ -0,0 +1,415 @@
...
+	hci_vs_evt_qos_conn_event_report_t *evt;
+
+	code = net_buf_simple_pull_u8(buf);
+	if (code != HCI_VS_SUBEVENT_CODE_QOS_CONN_EVENT_REPORT) {
+		return false;
+	}
+
+	evt = (void *)buf->data;
+	if ((evt->event_counter & 0x3FF) == 0) {
+		printk("QoS conn event reports: "
```
Comment (rugeGerritsen) [inline samples/bluetooth/llpm/src/main.c]: We only print the content of one report:
```suggestion
		printk("QoS conn event report: "
```
Maybe we should print the event counter as well?

File: samples/bluetooth/llpm/README.rst [hunk 28]
```diff
@@ -0,0 +1,313 @@
...
+
+       Transmission Latency: 1235 (us)
+
+.. msc::
+   hscale = "1.3";
+   Master,Slave;
+   Master<<=>>Slave         [label="Connected"];
+   Master<<=>>Slave         [label="Discovered BT_UUID_LATENCY SERVICE"];
+   Slave note Slave         [label="Press any key to start measuring transmission latency"];
+   Slave note Slave         [label="Read present timestamp: s1"];
```
Comment (rugeGerritsen) [inline samples/bluetooth/llpm/README.rst]: Should we use "Read current timestamp" instead of "Read present timestamp" instead? Similar for other occurrences in this MSC

File: samples/bluetooth/llpm/README.rst [hunk 29]
```diff
@@ -0,0 +1,313 @@
...
+The Bluetooth Low Latency Packet Mode (LLPM) sample uses the :ref:`latency_readme` to showcase the LLPM proprietary Bluetooth extension from Nordic Semiconductor.
+You can use it to determine the transmission latency of LLPM-enabled connections, or to compare with different connection parameters and check their influence on the results.
+
+
+Overview
+********
+
+The LLPM is designed time-sensitive transmissions, in which time synchronization is important and interface response time is critical for the user.
+For example, for reality headsets or gaming mouse and keyboard.
+
```
Comment (rugeGerritsen) [inline samples/bluetooth/llpm/README.rst]: I would insert the following here:

```suggestion
This sample determines the latency as shown in the MSC below::
```
(and then the MSC which is now placed further down)
In this way we make it clear how the latency is defined at an early point in the documentation.

File: samples/bluetooth/llpm/README.rst (hunk 24 above)
Comment (greg-fer) [reply samples/bluetooth/llpm/README.rst]:
```suggestion
Moreover, it uses the following nrfxlib library::
```
I'd say it like this. The repetition doesn't hurt, and we keep consistency with the previous sentence about Zephyr libs.

File: include/bluetooth/services/latency.rst [hunk 30]
```diff
@@ -0,0 +1,38 @@
...
+=============================================================
+
+Write
+   Write data to the characteristic and wait for a response to compute the time spent.
+
+
+API documentation
+*****************
+
+| Header file: :file:`include/latency.h`
```
Comment (rugeGerritsen) [inline include/bluetooth/services/latency.rst]: @greg-fer , does this work? I would expect we should use `include/bluetooth/services/latency.h`

File: samples/bluetooth/llpm/README.rst [hunk 31]
```diff
@@ -0,0 +1,313 @@
...
+   * - Physical layer (PHY)
+     - LE 2M PHY
+
+
+Requirements
+************
+
+* Two of the following nRF52-series development kit boards:
+
+  * PCA10040
```
[the remark this answers is not in the document]
Comment (greg-fer) [reply samples/bluetooth/llpm/README.rst]: The boards are so different that I don't see a way to rephrase this in a PCA-agnostic way. To my understanding, the samples will be checked every time for each new board, so the documentation will have to be updated at that stage. On the other hand, omitting this information from the sample documentation will create confusion.

File: samples/bluetooth/llpm/README.rst [hunk 32]
```diff
@@ -0,0 +1,313 @@
...
+Transmission latency
+   The definition of the latency used in this example counts the time interval from the sender's application to the GATT service of the receiver.
+   It demonstrates the performance of an LLPM-enabled connection that the receiver will receive the data approximately every 1 ms.
+
+GATT Latency Service
+   To measure the transmission latency from application layers, a GATT Latency service `BT_UUID_LATENCY` is included to compute the time spent.
+   When the sender writes its timestamp to the `BT_UUID_LATENC_CHAR` characteristic of the receiver, the Latency service of the receiver will automatically reply back.
+   Whenever the sender receives a response, it will use its present time and the corresponding timestamp written before to estimate the round-trip time (RTT) of a writing characteristic procedure (see `Bluetooth Core Specification`_: Vol 3, Part F, 3.4.5 Writing attributes).
+
+.. list-table:: GATT Attributes
```
Comment (rugeGerritsen) [inline samples/bluetooth/llpm/README.rst]: Should the service definition table for the latency service be moved to the latency service documentation? I believe it should be removed from the sample documentation

File: samples/bluetooth/llpm/README.rst (hunk 24 above)
Comment (rugeGerritsen) [reply samples/bluetooth/llpm/README.rst]: Should it rather be moved right below "This sample uses the following |NCS| libraries:"

File: include/bluetooth/services/latency.rst (hunk 30 above)
Comment (greg-fer) [reply include/bluetooth/services/latency.rst]: You're right, the path needs to be updated.

File: samples/bluetooth/llpm/README.rst (hunk 31 above)
[the remark this answers is not in the document]
Comment (rugeGerritsen) [reply samples/bluetooth/llpm/README.rst]: What about: 
Two nRF52-series development kit boards, e.g.:
?

File: samples/bluetooth/llpm/README.rst [hunk 33]
```diff
@@ -0,0 +1,313 @@
...
+     - UUID
+     - Property
+   * - Primary service
+     - BT_UUID_LATENCY
+     - Read only
+   * - Characteristic
+     - BT_UUID_LATENCY_CHAR
+     - BT_GATT_CHRC_WRITE, BT_GATT_PERM_WRITE
+
+This sample transmits data between two boards to measure the transmission latency in between.
```
Comment (rugeGerritsen) [inline samples/bluetooth/llpm/README.rst]: In the sample documentation we mention 3 times how the latency is computed. I would say that keeping only the MSC is sufficient.

File: samples/bluetooth/llpm/README.rst (hunk 24 above)
Comment (greg-fer) [reply samples/bluetooth/llpm/README.rst]: Need to check.

Comment (greg-fer) [reply samples/bluetooth/llpm/README.rst]: OK, checked. Please keep the nrfxlib line, but move it to line 277 (move the Zephyr sentence down). The nrfxlib sentence should read:
```
This sample uses the following `nrfxlib`_ libraries:

* :ref:`nrfxlib:ble_controller`
```
We don't need the file mentioned under this ref.

File: samples/bluetooth/llpm/README.rst [hunk 34]
```diff
@@ -0,0 +1,313 @@
...
+   hscale = "1.3";
+   Master,Slave;
+   Master<<=>>Slave         [label="Connected"];
+   Master<<=>>Slave         [label="Discovered BT_UUID_LATENCY SERVICE"];
+   Slave note Slave         [label="Press any key to start measuring transmission latency"];
+   Slave note Slave         [label="Read present timestamp: s1"];
+   Slave=>Master            [label="ATT Write Request (timestamp: s1)"];
+   Master>>Slave            [label="ATT Write response"];
+   Slave note Slave         [label="Read present timestamp: s2"];
+   Slave note Slave         [label="Slave Latency = (s2 - s1) / 2"];
```
Comment (rugeGerritsen) [inline samples/bluetooth/llpm/README.rst]: To avoid confusion, we should rather use the term "Latency, Slave" as "Slave Latency" is a BLE Spec feature. Similar for "Master Latency"

File: samples/bluetooth/llpm/README.rst (hunk 17 above)
Comment (rakons) [reply samples/bluetooth/llpm/README.rst]: Agree - I was looking for an information that this is not a Bluetooth standard.

File: samples/bluetooth/llpm/README.rst (hunk 34 above)
Comment (ryanjh) [reply samples/bluetooth/llpm/README.rst]: To avoid confusion, maybe we should just use "Latency = ...". Because we only divide the RTT in half, it is hard to judge that the latency belongs to master or slave. ??

File: samples/bluetooth/llpm/README.rst (hunk 31 above)
[the remark this answers is not in the document]
Comment (ryanjh) [reply samples/bluetooth/llpm/README.rst]: How about linking to the README of ble-controller (nrfxlib\ble_controller\README.rst)?

+--------------------------+--------+--------+--------+
|                          | S112   | S132   | S140   |
+==========================+========+========+========+
| Low Latency Packet mode  |        | X      | X      |
+--------------------------+--------+--------+--------+
| QoS Conn Event Reports   | X      | X      | X      |
+--------------------------+--------+--------+--------+

File: samples/bluetooth/llpm/src/main.c (hunk 18 above)
[the remark this answers is not in the document]
Comment (ryanjh) [reply samples/bluetooth/llpm/src/main.c]: Also the latency is printed in the terminal, the user doesn`t have to shift their focus away.

File: samples/bluetooth/llpm/README.rst (hunk 31 above)
[the remark this answers is not in the document]
Comment (joerchan) [reply samples/bluetooth/llpm/README.rst]: Leave as is. This is how it is done for all samples, so we should keep the pattern here as well.

File: samples/bluetooth/llpm/src/main.c [hunk 35]
```diff
@@ -0,0 +1,422 @@
...
+			printk("Enable LLPM short connection interval failed\n");
+			return;
+		}
+	}
+
+	printk("Press any key to start measuring transmission latency\n");
+	console_getchar();
+
+	/* Start sending the timestamp to its peer */
+	while (true) {
```
Comment (joerchan) [inline samples/bluetooth/llpm/src/main.c]: Break out of this loop when disconnected?
`while (default_conn) {`

File: samples/bluetooth/llpm/README.rst [hunk 36]
```diff
@@ -0,0 +1,306 @@
...
+
+
+Requirements
+************
+
+* Two of the following nRF52-series development kit boards:
+
+  * PCA10040
+  * PCA10056
+  * The others running LLPM-supported SoftDevice (see :ref:`nrfxlib:ble_controller` Proprietary feature support)
```
[the remark this answers is not in the document]
Comment (ryanjh) [reply samples/bluetooth/llpm/README.rst]: Per discussion with @greg-fer, I will rephrase it to BLE Controller variants in the document.
